quic/s2n-quic-core/src/buffer/receive_buffer.rs

```diff
@@ -124,16 +160,57 @@ impl ReceiveBuffer {
     #[inline]
     pub fn write_at(&mut self, offset: VarInt, data: &[u8]) -> Result<(), ReceiveBufferError> {
         // create a request
-        let request = Request::new(offset, data)?;
+        let request = Request::new(offset, data, false)?;
+        self.write_request(request)?;
+        Ok(())
+    }
+
+    /// Pushes a slice at a certain offset, which is the end of the buffer
+    #[inline]
+    pub fn write_at_fin(&mut self, offset: VarInt, data: &[u8]) -> Result<(), ReceiveBufferError> {
+        // create a request
+        let request = Request::new(offset, data, true)?;
 
+        // compute the final offset for the fin request
+        let final_offset = request.end_exclusive();
+
+        // make sure if we previously saw a final size that they still match
+        if self.final_offset != UNKNOWN_FINAL_SIZE && self.final_offset != final_offset {
+            return Err(ReceiveBufferError::InvalidFin);
```

Comment:
What happens when this error occurs, does the connection close? Is there an RFC citation that could go here and in the next check?

Reply:
Yes the connection closes. I can add a citation.